quickwit/quickwit-ingest/src/ingest_v2/routing_table.rs

```diff
@@ -0,0 +1,835 @@
+
+        if num_shards > local_shards.len() + remote_shards.len() {
+            warn!("input shards should not contain closed shards or duplicates");
+        }
+
+        Self {
+            index_uid,
+            source_id,
+            local_shards,
+            local_round_robin_idx: AtomicUsize::default(),
+            remote_shards,
+            remote_round_robin_idx: AtomicUsize::default(),
+        }
+    }
+
+    fn empty(index_uid: IndexUid, source_id: SourceId) -> Self {
+        Self {
+            index_uid,
+            source_id,
+            local_shards: Vec::new(),
+            local_round_robin_idx: AtomicUsize::default(),
+            remote_shards: Vec::new(),
+            remote_round_robin_idx: AtomicUsize::default(),
+        }
+    }
+
+    /// Returns `true` if at least one shard in the table entry is open and has a leader available.
+    /// As it goes through the list of shards in the entry, it populates `closed_shard_ids` and
+    /// `unavailable_leaders` with the shard IDs of the closed shards and the node ID of the
+    /// unavailable ingesters encountered along the way.
+    pub fn has_open_shards(
+        &self,
+        closed_shard_ids: &mut Vec<ShardId>,
+        ingester_pool: &IngesterPool,
+        unavailable_leaders: &mut HashSet<NodeId>,
+    ) -> bool {
+        for shards in [&self.local_shards, &self.remote_shards] {
+            for shard in shards {
+                if shard.shard_state.is_closed() {
+                    closed_shard_ids.push(shard.shard_id);
+                } else if shard.shard_state.is_open() {
+                    if ingester_pool.contains_key(&shard.leader_id) {
+                        return true;
+                    } else {
+                        let leader_id: NodeId = shard.leader_id.clone();
+                        unavailable_leaders.insert(leader_id);
+                    }
+                }
+            }
+        }
+        false
+    }
+
+    /// Returns the next open and available shard in the table entry in a round-robin fashion.
+    pub fn next_open_shard_round_robin(
+        &self,
+        ingester_pool: &IngesterPool,
+    ) -> Option<&RoutingEntry> {
+        for (shards, round_robin_idx) in [
+            (&self.local_shards, &self.local_round_robin_idx),
+            (&self.remote_shards, &self.remote_round_robin_idx),
+        ] {
+            if shards.is_empty() {
+                continue;
+            }
+            for _attempt in 0..shards.len() {
+                let shard_idx = round_robin_idx.fetch_add(1, Ordering::Relaxed);
+                let shard = &shards[shard_idx % shards.len()];
+
+                if shard.shard_state.is_open() && ingester_pool.contains_key(&shard.leader_id) {
+                    return Some(shard);
+                }
+            }
+        }
+        None
+    }
+
+    /// Inserts the open shards the routing table is not aware of.
+    fn insert_open_shards(
```

Comment:
@fulmicoton, this is the function I added to the routing table in this PR. This is where you should focus your review.